Tests/NIOTests/ByteBufferTest.swift

```diff
@@ -1704,6 +1704,37 @@ class ByteBufferTest: XCTestCase {
         XCTAssertEqual(0xcc, slice.getInteger(at: 0, as: UInt8.self))
         XCTAssertEqual(0xdd, slice.getInteger(at: slice.writerIndex - 1, as: UInt8.self))
     }
+    
+    func testSliceOnSliceAfterHitting16MBMark() {
+        // This test ensures that a slice will get a new backing storage if its start is more than
+        // 16MiB after the originating backing storage.
+        
+        // create a buffer with (16MiB - 1byte) + 2 bytes = 16MiB + 1 byte
+        let inputBufferLength = Int(_UInt24.max) + 2
+        var inputBuffer = ByteBuffer(repeating: 1, count: inputBufferLength)
+        // read a small slice from the inputBuffer, to create an offset of eight bytes
+        XCTAssertNotNil(inputBuffer.readInteger(as: UInt64.self))
+        
+        // read the remaining bytes into a new slice (this will have a length of 16MiB - 7Bbytes)
+        let remainingSliceLength = inputBufferLength - MemoryLayout<UInt64>.size
+        var remainingSlice = inputBuffer.readSlice(length: remainingSliceLength)!
+        
+        let finalSliceLength = 1
+        // let's create a new buffer that uses all but one byte
+        XCTAssertNotNil(remainingSlice.readSlice(length: remainingSliceLength - finalSliceLength))
+        
+        // there should only be one byte left.
+        XCTAssertEqual(remainingSlice.readableBytes, finalSliceLength)
+        
+        // the upper bound is exactly one byte above the 16MiB threshold. For this reason a slice
+        // of the last byte, will need to get a new backing storage.
+        XCTAssertEqual(remainingSlice._slice.upperBound, UInt32(_UInt24.max) + 2)
+        let finalSlice = remainingSlice.readSlice(length: finalSliceLength)
+        XCTAssertEqual(finalSlice?.storageCapacity, 1)
+        XCTAssertEqual(finalSlice?._slice.lowerBound, 0)
+        
```

Comment:
we should additionally assert
- that the slices/values read have the correct content (`1`s, `2`s, and `3`s or so as suggested above)
- that the underlying storage has indeed changed. There's a (test only) property called `buffer.storageIdentifier` or so which stays the same if nothing got reallocated/copied and changes if it has